rust/src/smb/detect.rs

```diff
@@ -201,3 +206,430 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
+                    return Err(());
+                }
+            }
+        }
+    }
+    return Ok(SmbCmdValidCodes::new(cmd_codes1, cmd_codes2));
+}
+
+fn gen_smb2_command_names() -> HashMap<String, u16> {
+    let commands = [
+        smb2::SMB2_COMMAND_NEGOTIATE_PROTOCOL,
+        smb2::SMB2_COMMAND_SESSION_SETUP,
+        smb2::SMB2_COMMAND_SESSION_LOGOFF,
+        smb2::SMB2_COMMAND_TREE_CONNECT,
+        smb2::SMB2_COMMAND_TREE_DISCONNECT,
+        smb2::SMB2_COMMAND_CREATE,
+        smb2::SMB2_COMMAND_CLOSE,
+        smb2::SMB2_COMMAND_FLUSH,
+        smb2::SMB2_COMMAND_READ,
+        smb2::SMB2_COMMAND_WRITE,
+        smb2::SMB2_COMMAND_LOCK,
+        smb2::SMB2_COMMAND_IOCTL,
+        smb2::SMB2_COMMAND_CANCEL,
+        smb2::SMB2_COMMAND_KEEPALIVE,
+        smb2::SMB2_COMMAND_FIND,
+        smb2::SMB2_COMMAND_CHANGE_NOTIFY,
+        smb2::SMB2_COMMAND_GET_INFO,
+        smb2::SMB2_COMMAND_SET_INFO,
+        smb2::SMB2_COMMAND_OPLOCK_BREAK,
+    ];
+    let mut cmd_names2 = HashMap::new();
+
+    for cmd in commands {
+        cmd_names2.insert(smb2::smb2_command_string(cmd), cmd);
+    }
+
+    cmd_names2.insert("negotiate".into(), smb2::SMB2_COMMAND_NEGOTIATE_PROTOCOL);
+    cmd_names2.insert(
+        "smb2_command_session_setup".into(),
+        smb2::SMB2_COMMAND_SESSION_SETUP,
+    );
+    cmd_names2.insert("session_setup".into(), smb2::SMB2_COMMAND_SESSION_SETUP);
+    cmd_names2.insert("logoff".into(), smb2::SMB2_COMMAND_SESSION_LOGOFF);
+    cmd_names2.insert("tree_connect".into(), smb2::SMB2_COMMAND_TREE_CONNECT);
+    cmd_names2.insert("tree_disconnect".into(), smb2::SMB2_COMMAND_TREE_DISCONNECT);
+    cmd_names2.insert("create".into(), smb2::SMB2_COMMAND_CREATE);
+    cmd_names2.insert("close".into(), smb2::SMB2_COMMAND_CLOSE);
+    cmd_names2.insert("flush".into(), smb2::SMB2_COMMAND_FLUSH);
+    cmd_names2.insert("read".into(), smb2::SMB2_COMMAND_READ);
+    cmd_names2.insert("write".into(), smb2::SMB2_COMMAND_WRITE);
+    cmd_names2.insert("lock".into(), smb2::SMB2_COMMAND_LOCK);
+    cmd_names2.insert("ioctl".into(), smb2::SMB2_COMMAND_IOCTL);
+    cmd_names2.insert("cancel".into(), smb2::SMB2_COMMAND_CANCEL);
+    cmd_names2.insert("echo".into(), smb2::SMB2_COMMAND_KEEPALIVE);
+    cmd_names2.insert("keep_alive".into(), smb2::SMB2_COMMAND_KEEPALIVE);
+    cmd_names2.insert("find".into(), smb2::SMB2_COMMAND_FIND);
+    cmd_names2.insert("query_directory".into(), smb2::SMB2_COMMAND_FIND);
+    cmd_names2.insert("change_notify".into(), smb2::SMB2_COMMAND_CHANGE_NOTIFY);
+    cmd_names2.insert("get_info".into(), smb2::SMB2_COMMAND_GET_INFO);
+    cmd_names2.insert("query_info".into(), smb2::SMB2_COMMAND_GET_INFO);
+    cmd_names2.insert("set_info".into(), smb2::SMB2_COMMAND_SET_INFO);
+    cmd_names2.insert("oplock_break".into(), smb2::SMB2_COMMAND_OPLOCK_BREAK);
+
+    return cmd_names2;
+}
+
+fn gen_smb1_command_names() -> HashMap<String, u16> {
+    let commands = [
+        smb1::SMB1_COMMAND_CREATE_DIRECTORY,
```

Comment:
Is there a more idiomatic way ?
cc @jasonish we want to iterate over all known values of an enumeration

Reply:
These code command constants are not part of an enumeration, but are the defined in the smb1 module.

On the other hand, there is no idiomatic way to iterate over values of enumeration since those couldn't be constants (https://stackoverflow.com/a/21373995)